rust/src/hdfs/block_writer.rs

```diff
@@ -362,11 +379,49 @@ impl ReplicatedBlockWriter {
 
         debug!("Recovering block writer");
 
+        let mut existing_block = self.block.clone();
+        let mut exclude_nodes = vec![];
+        for failed_node in &failed_nodes {
+            exclude_nodes.push(existing_block.locs[*failed_node].clone());
+
+            existing_block.locs.remove(*failed_node);
+            existing_block.storage_i_ds.remove(*failed_node);
+            existing_block.storage_types.remove(*failed_node);
+        }
         let mut new_block = self.block.clone();
-        for failed_node in failed_nodes {
-            new_block.locs.remove(failed_node);
-            new_block.storage_i_ds.remove(failed_node);
-            new_block.storage_types.remove(failed_node);
+
+        let should_replace = self.replace_datanode.should_replace(
+                new_block.locs.len() as u32,
```

Comment:
I think this should be the target replication of the file, not the current number of blocks? If we have a replication of 3 and have already failed to one of the replicas, this would only be 2, when we really want to still use 3 here I think? Might require passing the file status into the block writer

Reply:
I'm not sure if the variable naming is misleading. Self.block should be used here instead of new_block, although new_block is just copied from self.block and its content is the same as self.block. I have fixed it in the latest code.
As far as I understand, the length of self.block.locs will not change, so using self.block.locs should be ok, right?

Reply:
It's a very minor edge with the way the math works out, but basically the scenario I'm thinking of is:
- You are using the default datanode replacement policy
- You create a file with a replication of 4
- One of the datanodes fails, because the default policy will only replace a datanode when `4 - (4 / 2)` datanodes fail
- `self.block` now contains just three datanodes as the failed datanode was removed
- Another datanode fails during the write. The default policy _should_ replace the datanode now, but the policy thinks the replication is 3 because `self.blocks` has 3 entries, when in reality the target replication of the file is 4

That case doesn't happen with the default replication of 3 since `3/2 == 2/2`, but might still be good to try to handle it properly

Reply:
makes sense.